File: src/collectors/bridges.js
'use strict';
/**
 * Bridges collector.
 *
 *   /interface/bridge        the bridges themselves — STP mode, VLAN filtering,
 *                            IGMP snooping, priority, ageing, MAC, MTU
 *   /interface/bridge/port   each port's STP role, edge/learn/horizon and PVID
 *   /interface/bridge/host   the learned MAC table
 *
 * /interface/bridge/vlan IS NOT READ. The VLAN membership table belongs to the
 * VLANs page, which already fetches it: that table is about which VLANs exist
 * and where they are tagged, not about how a bridge is configured. Reading it
 * here as well would be a second copy of the same rows, on a second poll loop,
 * for a card that no longer exists.
 *
 * WHAT WAS ALREADY AVAILABLE, AND WHY THIS IS STILL NEW. Bridges show up in
 * interfaceStatus today as ordinary interfaces with type 'bridge' — a name, a
 * running flag and rates, and nothing else. Every field that makes a bridge a
 * bridge (protocol-mode, vlan-filtering, igmp-snooping, port roles) is
 * unreadable from there. vlans.js does read the port and vlan menus, but with a
 * proplist that omits role/edge/learn/horizon because the VLAN page has no use
 * for them.
 *
 * RATES ARE BORROWED, NOT FETCHED, the way vlans.js borrows them: ifStatus
 * already computes rxMbps/txMbps for every interface, bridges included, so a
 * bridge's throughput costs no extra router I/O. Fields are projected BY NAME
 * rather than spread, so nothing from the interface payload (IP addresses, MAC
 * lists, error counters) reaches a page whose permission is `bridges` rather
 * than `interfaces`.
 *
 * THE HOST TABLE IS CAPPED. 64 entries on the router this was written against,
 * but a switch with a few hundred clients will happily return thousands, every
 * poll, over the socket. The cap is applied here and the true total travels
 * with it so the page can say "showing 500 of 2431" rather than quietly lying.
 */

const { clampPoll, createPollLoop, createListenRefresh } = require('./util');

const BRIDGE_CMD = ['/interface/bridge/print',
                    '=.proplist=.id,name,protocol-mode,vlan-filtering,igmp-snooping,dhcp-snooping,' +
                    'fast-forward,priority,ageing-time,mac-address,actual-mtu,mtu,running,disabled,comment'];
const PORT_CMD   = ['/interface/bridge/port/print',
                    '=.proplist=.id,bridge,interface,pvid,role,edge,learn,horizon,path-cost,' +
                    'frame-types,disabled,inactive,dynamic'];
const HOST_CMD   = ['/interface/bridge/host/print',
                    '=.proplist=mac-address,on-interface,bridge,vid,dynamic,local,external,age'];
const LISTEN_CMD = '/interface/bridge/port/listen';
const HOST_CAP   = 500;
// In stream mode the bridge and port tables are re-read on a listen event; this
// is the safety net, so a missed event cannot strand the page on stale config.
const CONFIG_EVERY = 12;

const _bool = (v) => v === true || v === 'true' || v === 'yes';
const _num  = (v) => { const n = Number(v); return Number.isFinite(n) ? n : null; };

/**
 * Join the three tables into one view.
 *
 * `ifPayload` is interfaceStatus's lastPayload or null — null must yield rates
 * of null, never 0, because "the router did not report this" and "this bridge
 * is idle" have to stay tellable apart.
 */
function buildBridgeRows(bridgeRows, portRows, hostRows, ifPayload) {
  const rates = new Map();
  const ifaces = (ifPayload && Array.isArray(ifPayload.interfaces)) ? ifPayload.interfaces : null;
  if (ifaces) for (const i of ifaces) if (i && i.name) rates.set(i.name, i);

  const ports = [];
  for (const r of portRows || []) {
    if (!r || !r.interface) continue;              // also drops {undefined:''}
    ports.push({
      // The row id, so the page can open a port in the edit form.
      id:         r['.id'] || '',
      bridge:     r.bridge || '',
      interface:  String(r.interface),
      pvid:       _num(r.pvid),
      // Absent on a bridge running protocol-mode=none: there are no STP roles
      // to report, which is not the same as a port with an unknown role.
      role:       r.role || '',
      edge:       r.edge || '',
      learn:      r.learn || '',
      horizon:    r.horizon || '',
      pathCost:   _num(r['path-cost']),
      frameTypes: r['frame-types'] || '',
      disabled:   _bool(r.disabled),
      inactive:   _bool(r.inactive),
      dynamic:    _bool(r.dynamic),
    });
  }

  const portsByBridge = new Map();
  for (const p of ports) {
    if (!p.bridge) continue;
    portsByBridge.set(p.bridge, (portsByBridge.get(p.bridge) || 0) + 1);
  }

  const bridges = [];
  for (const r of bridgeRows || []) {
    if (!r || !r.name) continue;
    const live = rates.get(r.name);
    bridges.push({
      // The row id, so the page can open a bridge in the edit form.
      id:            r['.id'] || '',
      name:          String(r.name),
      protocolMode:  r['protocol-mode'] || '',
      vlanFiltering: _bool(r['vlan-filtering']),
      igmpSnooping:  _bool(r['igmp-snooping']),
      dhcpSnooping:  _bool(r['dhcp-snooping']),
      fastForward:   _bool(r['fast-forward']),
      priority:      r.priority || '',
      ageingTime:    r['ageing-time'] || '',
      macAddress:    r['mac-address'] || '',
      mtu:           _num(r['actual-mtu']) !== null ? _num(r['actual-mtu']) : _num(r.mtu),
      running:       _bool(r.running),
      disabled:      _bool(r.disabled),
      comment:       r.comment || '',
      portCount:     portsByBridge.get(String(r.name)) || 0,
      // null, never 0: "not reported" and "idle" must stay distinguishable.
      rxMbps:        live && typeof live.rxMbps === 'number' ? live.rxMbps : null,
      txMbps:        live && typeof live.txMbps === 'number' ? live.txMbps : null,
    });
  }
  bridges.sort((a, b) => a.name.localeCompare(b.name));

  const allHosts = [];
  for (const r of hostRows || []) {
    if (!r || !r['mac-address']) continue;
    allHosts.push({
      mac:         String(r['mac-address']),
      onInterface: r['on-interface'] || '',
      bridge:      r.bridge || '',
      vid:         _num(r.vid),
      dynamic:     _bool(r.dynamic),
      local:       _bool(r.local),
      external:    _bool(r.external),
      age:         r.age || '',
    });
  }
  // Learned entries first: a table truncated at the cap should drop the router's
  // own port MACs before it drops a client somebody is looking for.
  allHosts.sort((a, b) => (a.local === b.local)
    ? a.mac.localeCompare(b.mac)
    : (a.local ? 1 : -1));

  return {
    bridges,
    ports,
    hosts:      allHosts.slice(0, HOST_CAP),
    hostTotal:  allHosts.length,
    hostCap:    HOST_CAP,
    ratesAvailable: !!ifaces,
  };
}

class BridgesCollector {
  constructor({ ros, io, state, pollMs, ifStatus, streamMode }) {
    this.ros    = ros;
    this.io     = io;
    this.state  = state;
    this.pollMs = clampPoll(pollMs, 5000, 60000, 2000);
    this._lbl   = ros.routerLabel ? `[${ros.routerLabel}][bridges]` : '[bridges]';

    // DISABLEABLE, so this may be a null-collector stub with lastPayload null.
    // Guarded at every read.
    this.ifStatus = ifStatus || null;

    // Delivery switch (#105). Both modes emit at pollMs — the rates come from
    // ifStatus in memory and have to keep moving either way — but stream mode
    // re-reads the bridge and port tables only when the router says they
    // changed, and poll mode holds no channel open at all.
    this.streamMode = streamMode !== false;
    this._poll   = createPollLoop(() => this._tick(), () => this.pollMs);
    this._listen = createListenRefresh({
      ros, cmd: LISTEN_CMD, label: this._lbl,
      onEvent: () => { this._dirty = true; this._tick().catch(() => {}); },
    });
    // The first tick always reads; after that stream mode waits to be told.
    this._dirty  = true;
    this._ticks  = 0;
    this._cfg    = { bridgeRows: [], portRows: [] };
    this._lastFp = '';
    // undefined = unprobed, false = this router has no such menu, stop asking.
    this._bridgeAvailable = undefined;
    this._portAvailable   = undefined;
    this._hostAvailable   = undefined;
    this.lastPayload = null;
  }

  async _read(cmd, flag) {
    if (this[flag] === false) return [];
    try {
      const rows = await this.ros.write(cmd[0], [cmd[1]]);
      this[flag] = true;
      return (rows || []).filter(r => r && Object.keys(r).length);
    } catch (e) {
      const msg = String((e && e.message) || e).toLowerCase();
      if (msg.includes('no such') || msg.includes('unknown command')) this[flag] = false;
      // The host table is the one menu a read-only API user can be denied while
      // the rest still answers; topology.js latches the same way rather than
      // asking forever.
      else if (msg.includes('not enough permissions') || msg.includes('permission denied')) this[flag] = false;
      else this.state.lastBridgesErr = e && e.message ? e.message : String(e);
      return [];
    }
  }

  /**
   * Re-read now, after a write, so the page shows what the router did.
   *
   * Sets `_dirty` rather than reaching past _tick(): that flag already means
   * "config changed, read it on this tick", which is exactly the situation a
   * write creates.
   */
  async refreshNow() {
    if (!this.ros.connected) return;
    this._dirty = true;
    await this._tick();
  }

  async _tick() {
    if (!this.ros.connected) return;

    // Config: every tick in poll mode; on a listen event or the safety interval
    // in stream mode. Hosts are volatile — a MAC is learned or ages out with no
    // config event behind it — so they are read every tick in both modes.
    const readConfig = !this.streamMode || this._dirty || (this._ticks % CONFIG_EVERY === 0);
    this._ticks++;
    if (readConfig) {
      const [bridgeRows, portRows] = await Promise.all([
        this._read(BRIDGE_CMD, '_bridgeAvailable'),
        this._read(PORT_CMD,   '_portAvailable'),
      ]);
      this._cfg = { bridgeRows, portRows };
      this._dirty = false;
    }
    const hostRows = await this._read(HOST_CMD, '_hostAvailable');

    const built = buildBridgeRows(this._cfg.bridgeRows, this._cfg.portRows, hostRows,
                                  this.ifStatus && this.ifStatus.lastPayload);
    const payload = {
      ts: Date.now(), pollMs: this.pollMs,
      ...built,
      // So the page can say "this router has no bridges" rather than showing an
      // empty table, which reads as a failure.
      available:      this._bridgeAvailable !== false,
      hostsAvailable: this._hostAvailable !== false,
    };
    // Assigned unconditionally: sendInitialState replays it, so a socket that
    // connects during a quiet spell must still get the current view.
    this.lastPayload = payload;
    this.state.lastBridgesTs = payload.ts;

    const fp = JSON.stringify({
      b: built.bridges.map(b => [b.name, b.protocolMode, b.vlanFiltering, b.igmpSnooping,
                                 b.running, b.portCount, b.rxMbps, b.txMbps]),
      p: built.ports.map(p => [p.interface, p.bridge, p.pvid, p.role, p.inactive, p.disabled]),
      h: built.hostTotal,
    });
    if (fp === this._lastFp) return;
    this._lastFp = fp;
    this.io.to('page-bridges').emit('bridges:update', payload);
  }

  _startDelivery() {
    this._poll.start();
    if (this.streamMode) this._listen.start();
  }

  _stopDelivery() {
    this._poll.stop();
    this._listen.stop();
  }

  async start() {
    if (this.ros.connected) await this._tick();
    this._startDelivery();
    this.ros.on('close', () => this._stopDelivery());
    this.ros.on('connected', async () => {
      this._stopDelivery();
      // A reconnect may be a different router, so every latch and the config
      // cadence reset with it.
      this._lastFp = '';
      this._dirty  = true;
      this._ticks  = 0;
      this._bridgeAvailable = this._portAvailable = this._hostAvailable = undefined;
      await this._tick();
      this._startDelivery();
    });
  }

  suspend() { this._stopDelivery(); }
  resume()  { if (this.ros.connected) this._startDelivery(); }

  stop() { this._stopDelivery(); this._lastFp = ''; }
}

BridgesCollector.buildBridgeRows = buildBridgeRows;
BridgesCollector.HOST_CAP        = HOST_CAP;
module.exports = BridgesCollector;

File: internal/collect/bridges.go
package collect

// Bridges collector — the port of src/collectors/bridges.js.
//
//	/interface/bridge        the bridges themselves
//	/interface/bridge/port   each port's STP role, edge/learn/horizon and PVID
//	/interface/bridge/host   the learned MAC table
//
// `/interface/bridge/vlan` IS NOT READ. That table belongs to the VLANs page,
// which already fetches it; reading it here would be a second copy of the same
// rows on a second poll loop.
//
// RATES ARE BORROWED, NOT FETCHED. interfaceStatus already computes rxMbps and
// txMbps for every interface, bridges included, so a bridge's throughput costs
// no extra router I/O. Fields are projected BY NAME rather than copied wholesale,
// so nothing from the interface payload — addresses, MAC lists, error counters —
// reaches a page whose permission is `bridges` rather than `interfaces`.
//
// THE HOST TABLE IS CAPPED. 64 entries on the router this was written against,
// but a switch with a few hundred clients returns thousands, every poll, over
// the socket. The cap is applied here and the true total travels with it, so the
// page can say "showing 500 of 2431" rather than quietly lying.

import (
	"encoding/json"
	"sort"
	"sync"
	"time"

	"mikrodash/internal/routeros"
)

var (
	bridgeCmd = routeros.Cmd{Path: "/interface/bridge/print", Args: []string{
		"=.proplist=.id,name,protocol-mode,vlan-filtering,igmp-snooping,dhcp-snooping," +
			"fast-forward,priority,ageing-time,mac-address,actual-mtu,mtu,running,disabled,comment"}}
	bridgePortCmd = routeros.Cmd{Path: "/interface/bridge/port/print", Args: []string{
		"=.proplist=.id,bridge,interface,pvid,role,edge,learn,horizon,path-cost," +
			"frame-types,disabled,inactive,dynamic"}}
	bridgeHostCmd = routeros.Cmd{Path: "/interface/bridge/host/print", Args: []string{
		"=.proplist=mac-address,on-interface,bridge,vid,dynamic,local,external,age"}}
)

const (
	bridgeHostCap = 500
	// The safety net: config is re-read every twelfth tick even when nothing
	// has said it changed, so a missed change cannot strand the page on stale
	// configuration.
	bridgeConfigEvery = 12
)

// Rate is one interface's throughput, as interfaceStatus reports it.
type Rate struct {
	RxMbps *float64
	TxMbps *float64
}

// RateSource supplies borrowed throughput.
//
// The bool is not redundant with an empty map. "interfaceStatus has not reported
// yet" and "it reported and this bridge is idle" must stay tellable apart — the
// page renders the first as an em dash and the second as 0.00, and conflating
// them makes a collector that has been switched off look like a quiet network.
type RateSource interface {
	Rates() (map[string]Rate, bool)
}

// Bridge is one row of the bridges table.
type Bridge struct {
	ID            string   `json:"id"`
	Name          string   `json:"name"`
	ProtocolMode  string   `json:"protocolMode"`
	VlanFiltering bool     `json:"vlanFiltering"`
	IgmpSnooping  bool     `json:"igmpSnooping"`
	DhcpSnooping  bool     `json:"dhcpSnooping"`
	FastForward   bool     `json:"fastForward"`
	Priority      string   `json:"priority"`
	AgeingTime    string   `json:"ageingTime"`
	MacAddress    string   `json:"macAddress"`
	MTU           *float64 `json:"mtu"`
	Running       bool     `json:"running"`
	Disabled      bool     `json:"disabled"`
	Comment       string   `json:"comment"`
	PortCount     int      `json:"portCount"`
	// null, never 0 — see RateSource.
	RxMbps *float64 `json:"rxMbps"`
	TxMbps *float64 `json:"txMbps"`
}

// BridgePort is one row of the ports table.
type BridgePort struct {
	ID        string   `json:"id"`
	Bridge    string   `json:"bridge"`
	Interface string   `json:"interface"`
	Pvid      *float64 `json:"pvid"`
	// Absent on a bridge running protocol-mode=none: there are no STP roles to
	// report, which is not the same as a port with an unknown role.
	Role       string   `json:"role"`
	Edge       string   `json:"edge"`
	Learn      string   `json:"learn"`
	Horizon    string   `json:"horizon"`
	PathCost   *float64 `json:"pathCost"`
	FrameTypes string   `json:"frameTypes"`
	Disabled   bool     `json:"disabled"`
	Inactive   bool     `json:"inactive"`
	Dynamic    bool     `json:"dynamic"`
}

// BridgeHost is one learned MAC.
type BridgeHost struct {
	Mac         string   `json:"mac"`
	OnInterface string   `json:"onInterface"`
	Bridge      string   `json:"bridge"`
	Vid         *float64 `json:"vid"`
	Dynamic     bool     `json:"dynamic"`
	Local       bool     `json:"local"`
	External    bool     `json:"external"`
	Age         string   `json:"age"`
}

// BridgesPayload is the bridges:update body. Field order matches the Node
// object literal so the emitted JSON reads the same.
type BridgesPayload struct {
	TS             int64        `json:"ts"`
	PollMs         int          `json:"pollMs"`
	Bridges        []Bridge     `json:"bridges"`
	Ports          []BridgePort `json:"ports"`
	Hosts          []BridgeHost `json:"hosts"`
	HostTotal      int          `json:"hostTotal"`
	HostCap        int          `json:"hostCap"`
	RatesAvailable bool         `json:"ratesAvailable"`
	// So the page can say "this router has no bridges" rather than showing an
	// empty table, which reads as a failure.
	Available      bool `json:"available"`
	HostsAvailable bool `json:"hostsAvailable"`
}

type bridgeBuilt struct {
	bridges        []Bridge
	ports          []BridgePort
	hosts          []BridgeHost
	hostTotal      int
	ratesAvailable bool
}

// BuildBridgeRows joins the three tables into one view.
func BuildBridgeRows(bridgeRows, portRows, hostRows []routeros.Reply, rates RateSource) bridgeBuilt {
	var byName map[string]Rate
	ratesAvailable := false
	if rates != nil {
		byName, ratesAvailable = rates.Rates()
	}

	ports := []BridgePort{}
	for _, r := range portRows {
		if r["interface"] == "" {
			continue // also drops the {undefined:''} row RouterOS can send
		}
		ports = append(ports, BridgePort{
			ID: r[".id"], Bridge: r["bridge"], Interface: r["interface"],
			Pvid: numOf(r, "pvid"), Role: r["role"], Edge: r["edge"],
			Learn: r["learn"], Horizon: r["horizon"], PathCost: numOf(r, "path-cost"),
			FrameTypes: r["frame-types"], Disabled: boolOf(r["disabled"]),
			Inactive: boolOf(r["inactive"]), Dynamic: boolOf(r["dynamic"]),
		})
	}

	portsByBridge := map[string]int{}
	for _, p := range ports {
		if p.Bridge != "" {
			portsByBridge[p.Bridge]++
		}
	}

	bridges := []Bridge{}
	for _, r := range bridgeRows {
		if r["name"] == "" {
			continue
		}
		// actual-mtu is what the interface negotiated; mtu is what it was
		// configured with. The first is the truth when both are present.
		mtu := numOf(r, "actual-mtu")
		if mtu == nil {
			mtu = numOf(r, "mtu")
		}
		b := Bridge{
			ID: r[".id"], Name: r["name"], ProtocolMode: r["protocol-mode"],
			VlanFiltering: boolOf(r["vlan-filtering"]), IgmpSnooping: boolOf(r["igmp-snooping"]),
			DhcpSnooping: boolOf(r["dhcp-snooping"]), FastForward: boolOf(r["fast-forward"]),
			Priority: r["priority"], AgeingTime: r["ageing-time"], MacAddress: r["mac-address"],
			MTU: mtu, Running: boolOf(r["running"]), Disabled: boolOf(r["disabled"]),
			Comment: r["comment"], PortCount: portsByBridge[r["name"]],
		}
		if live, ok := byName[b.Name]; ok {
			b.RxMbps, b.TxMbps = live.RxMbps, live.TxMbps
		}
		bridges = append(bridges, b)
	}
	sort.SliceStable(bridges, func(i, j int) bool {
		return Collate(bridges[i].Name, bridges[j].Name) < 0
	})

	all := []BridgeHost{}
	for _, r := range hostRows {
		if r["mac-address"] == "" {
			continue
		}
		all = append(all, BridgeHost{
			Mac: r["mac-address"], OnInterface: r["on-interface"], Bridge: r["bridge"],
			Vid: numOf(r, "vid"), Dynamic: boolOf(r["dynamic"]),
			Local: boolOf(r["local"]), External: boolOf(r["external"]), Age: r["age"],
		})
	}
	// Learned entries first: a table truncated at the cap should drop the
	// router's own port MACs before it drops a client somebody is looking for.
	sort.SliceStable(all, func(i, j int) bool {
		if all[i].Local == all[j].Local {
			return Collate(all[i].Mac, all[j].Mac) < 0
		}
		return !all[i].Local
	})

	hosts := all
	if len(hosts) > bridgeHostCap {
		hosts = hosts[:bridgeHostCap]
	}
	return bridgeBuilt{bridges, ports, hosts, len(all), ratesAvailable}
}

// Bridges is the collector.
type Bridges struct {
	ros    Reader
	emit   Emit
	rates  RateSource
	pollMs *pollInterval

	poll *pollLoop

	mu     sync.Mutex
	cfgB   []routeros.Reply
	cfgP   []routeros.Reply
	dirty  bool
	ticks  int
	lastFp string
	// nil = unprobed, false = this router has no such menu, stop asking.
	bridgeAvailable *bool
	portAvailable   *bool
	hostAvailable   *bool

	last    *BridgesPayload
	lastErr string
}

// NewBridges builds the collector. `rates` may be nil — the page degrades to no
// throughput column rather than not rendering, which is the same judgement
// src/collection.js makes by declaring no `requires` for this collector.
func NewBridges(ros Reader, emit Emit, rates RateSource, pollMs int) *Bridges {
	b := &Bridges{
		ros: ros, emit: emit, rates: rates,
		pollMs: newPollInterval(clampPoll(pollMs, 5000, 2000, 60000)),
		dirty:  true,
	}
	b.poll = newPollLoop(func() { b.Tick() }, func() time.Duration {
		return b.pollMs.duration()
	})
	return b
}

func (b *Bridges) read(cmd routeros.Cmd, avail **bool) []routeros.Reply {
	if *avail != nil && !**avail {
		return nil
	}
	rows, err := b.ros.Do(cmd)
	if err != nil {
		// The host table is the one menu a read-only API user can be denied
		// while the rest still answers, so a denial latches rather than being
		// asked for again every tick.
		if menuMissing(err) {
			no := false
			*avail = &no
		} else {
			b.lastErr = err.Error()
		}
		return nil
	}
	yes := true
	*avail = &yes
	out := make([]routeros.Reply, 0, len(rows))
	for _, r := range rows {
		if len(r) > 0 {
			out = append(out, r)
		}
	}
	return out
}

// Tick reads what is due this cycle and emits when something changed.
func (b *Bridges) Tick() {
	if !b.ros.Connected() {
		return
	}
	b.mu.Lock()
	defer b.mu.Unlock()

	// Config on the slow cadence; hosts every tick, because a MAC is learned or
	// ages out with no configuration change behind it.
	if b.dirty || b.ticks%bridgeConfigEvery == 0 {
		b.cfgB = b.read(bridgeCmd, &b.bridgeAvailable)
		b.cfgP = b.read(bridgePortCmd, &b.portAvailable)
		b.dirty = false
	}
	b.ticks++
	hostRows := b.read(bridgeHostCmd, &b.hostAvailable)

	built := BuildBridgeRows(b.cfgB, b.cfgP, hostRows, b.rates)
	payload := &BridgesPayload{
		TS: time.Now().UnixMilli(), PollMs: b.pollMs.ms(),
		Bridges: built.bridges, Ports: built.ports, Hosts: built.hosts,
		HostTotal: built.hostTotal, HostCap: bridgeHostCap,
		RatesAvailable: built.ratesAvailable,
		Available:      b.bridgeAvailable == nil || *b.bridgeAvailable,
		HostsAvailable: b.hostAvailable == nil || *b.hostAvailable,
	}
	// Assigned unconditionally: a socket that connects during a quiet spell is
	// replayed this, so it must be current even when nothing is emitted.
	b.last = payload

	// The whole rows, not a hand-picked tuple. The Node side learned this on the
	// DNS collector: a tuple omits a field the page renders, an edit to that
	// field recomputes an identical fingerprint, and the open page never hears
	// about a change that really landed. hostTotal rather than the hosts
	// themselves — the MAC table churns constantly and fingerprinting it would
	// emit every tick, which is what the fingerprint exists to prevent.
	fp, _ := json.Marshal(struct {
		B []Bridge     `json:"b"`
		P []BridgePort `json:"p"`
		H int          `json:"h"`
	}{built.bridges, built.ports, built.hostTotal})
	if string(fp) == b.lastFp {
		return
	}
	b.lastFp = string(fp)
	b.emit("page-bridges", "bridges:update", payload)
}

// Last is the most recent payload, replayed to a socket that has just opened
// the page so it is not blank for a whole poll interval.
func (b *Bridges) Last() *BridgesPayload {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.last
}

// RefreshNow re-reads immediately, after a write.
//
// Sets `dirty` rather than reaching past Tick: that flag already means "config
// changed, read it on this tick", which is exactly what a write creates.
func (b *Bridges) RefreshNow() {
	if !b.ros.Connected() {
		return
	}
	b.mu.Lock()
	b.dirty = true
	b.mu.Unlock()
	b.Tick()
}

func (b *Bridges) Start() {
	if b.ros.Connected() {
		b.Tick()
	}
	b.poll.start()
}

// Reconnected drops every latch: a reconnect may be a different build, so an
// "absent menu" decision taken against the old one must not persist.
func (b *Bridges) Reconnected() {
	b.poll.stop()
	b.mu.Lock()
	b.lastFp = ""
	b.dirty = true
	b.ticks = 0
	b.bridgeAvailable, b.portAvailable, b.hostAvailable = nil, nil, nil
	b.mu.Unlock()
	b.Tick()
	b.poll.start()
}

func (b *Bridges) Suspend() { b.poll.stop() }

func (b *Bridges) Resume() {
	if b.ros.Connected() {
		b.poll.start()
	}
}

func (b *Bridges) Stop() {
	b.poll.stop()
	b.mu.Lock()
	b.lastFp = ""
	b.mu.Unlock()
}

// SetPollMs applies a new poll period to a running collector.
// See `System.SetPollMs` for why both halves are needed.
func (b *Bridges) SetPollMs(ms int) {
	b.pollMs.set(ms)
	b.poll.retime()
}
